### PR TITLE
Forbidden non-number type cast reduce

### LINKED ISSUE #4011: query result error in norec mode
with enable pipeline or disable pipeline, the query results are both not same.
### Steps to reproduce the behavior (Required)
1. DROP DATABASE IF EXISTS n1;
CREATE DATABASE n1;
USE n1;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE t0 (c_0_0 BOOLEAN NULL ,c_0_1 DATETIME NULL ,c_0_2 DECIMAL(24, 18)  NOT NULL ,c_0_3 DECIMAL(21, 9)  NULL ,c_0_4 FLOAT NULL ,c_0_5 DOUBLE NULL ) UNIQUE KEY (c_0_0,c_0_1,c_0_2,c_0_3) DISTRIBUTED BY HASH (c_0_0,c_0_2) properties("replication_num"="3") ;
CREATE TABLE IF NOT EXISTS t1 (c_1_0 VARCHAR(1) NOT NULL ,c_1_1 DECIMAL(3, 0)  NULL ,c_1_2 VARCHAR(21) NULL ,c_1_3 BOOLEAN NOT NULL ,c_1_4 DATETIME NOT NULL ,c_1_5 LARGEINT NULL ,c_1_6 DATE NULL ,c_1_7 BOOLEAN NOT NULL ,c_1_8 INT NULL ,c_1_9 BOOLEAN NOT NULL ,c_1_10 VARCHAR(21) NOT NULL ,c_1_11 VARCHAR(31) NOT NULL ,c_1_12 VARCHAR(11) NOT NULL ,c_1_13 BOOLEAN NOT NULL ,c_1_14 CHAR(11) NULL ,c_1_15 DOUBLE NOT NULL ,c_1_16 TINYINT NULL ,c_1_17 SMALLINT NOT NULL ) DUPLICATE KEY (c_1_0,c_1_1,c_1_2,c_1_3,c_1_4,c_1_5,c_1_6,c_1_7,c_1_8,c_1_9,c_1_10,c_1_11,c_1_12) DISTRIBUTED BY HASH (c_1_2,c_1_11,c_1_12,c_1_1,c_1_4,c_1_3,c_1_5,c_1_9,c_1_8,c_1_7,c_1_10,c_1_0,c_1_6) properties("replication_num"="3") ;
INSERT INTO t0 VALUES (true, NULL, 8.47138E-13, 436.041685040, 0.5813459275403033, 0.2492299217531957), (false, "1970-01-01 08:00:00", 8.93200E-13, NULL, -9.65483623E8, 0.7897373691829063);
INSERT INTO t0 VALUES (true, "1970-01-01 08:00:00", 6.49661E-13, NULL, -1.471014101E9, 0.4792647976982001);
INSERT INTO t1 VALUES ("", 11, "SL", true, "1970-01-01 08:00:00", NULL, "1970-01-01", false, 9, true, "U9E", "A", "zM6PgI2Vy", true, "lfT", 0.04347456563216889, 6, 5), ("", 620, "", false, "1970-01-01 08:00:00", -1460101275, "1970-01-01", true, 6, true, "w", "Rmz9eqm4D", "", true, "Aj", 0.031502332921569676, 4, 6), ("", 35, "", true, "1970-01-01 08:00:00", 60918059, "1970-01-01", false, 21, false, "", "uJ0pj", "", true, "qd0", -1.395477296E9, 4, 13);
DELETE FROM t1 WHERE c_1_7=false AND c_1_6="1969-12-15" AND c_1_1=323 AND c_1_5=-1471014101 AND c_1_8=2086333941 AND c_1_11="8" AND c_1_9=true AND c_1_0="" AND c_1_4="1969-12-16 16:08:59";
INSERT INTO t0 VALUES (NULL, "1970-01-01 08:00:00", 9.56409E-13, 447.134950965, 0.6892508068682921, -1.460101275E9), (true, "1970-01-01 08:00:00", 9.81489E-13, 332.010971835, 1.908629271E9, 0.6746607012580524), (false, "1970-01-01 08:00:00", 1.9936E-14, 545.983837768, 0.9627388029048002, 0.788111180816273);
INSERT INTO t1 VALUES ("", NULL, "", true, "1970-01-01 08:00:00", -965483623, "1970-01-01", false, 7, false, "3", "o", "", false, "jG", 0.5813459275403033, 3, 10);
INSERT INTO t0 VALUES (false, "1970-01-01 08:00:00", 8.57041E-13, 764.081589991, 0.5210565647263977, 0.9902506377323703);
2. mysql> SELECT COUNT(*) FROM t0 WHERE CAST(CASE  WHEN (t0.c_0_2)  BETWEEN (CASE  WHEN CASE  WHEN true THEN true END THEN t0.c_0_2  WHEN true THEN t0.c_0_2 ELSE t0.c_0_2 END) AND (t0.c_0_2)  THEN t0.c_0_1  WHEN false THEN "1969-12-22 10:50:55" ELSE "1969-12-30 10:15:23" END AS BOOLEAN ) ;
+----------+
| count(*) |
+----------+
|        6 |
+----------+
mysql> SELECT SUM(count) FROM (SELECT CAST((CAST(CAST(CASE  WHEN (t0.c_0_2)  BETWEEN (CASE  WHEN CASE  WHEN true THEN true END THEN t0.c_0_2  WHEN true THEN t0.c_0_2 ELSE t0.c_0_2 END) AND (t0.c_0_2)  THEN t0.c_0_1  WHEN false THEN "1969-12-22 10:50:55" ELSE "1969-12-30 10:15:23" END AS BOOLEAN )  AS BOOLEAN) = true) AND (CAST(CAST(CASE  WHEN (t0.c_0_2)
 BETWEEN (CASE  WHEN CASE  WHEN true THEN true END THEN t0.c_0_2  WHEN true THEN t0.c_0_2 ELSE t0.c_0_2 END) AND (t0.c_0_2)  THEN t0.c_0_1  WHEN false THEN "1969-12-22 10:50:55"
ELSE "1969-12-30 10:15:23" END AS BOOLEAN )  AS BOOLEAN) IS NOT NULL) AS INT) as count FROM t0 ) t;
+------------+
| sum(count) |
+------------+
|          0 |
+------------+

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - main Release
 - mysql> select current_version();
+---------------------------------+
| current_version()               |
+---------------------------------+
| SQLANCER_02_RELEASE 3ff7605 |
+---------------------------------+

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4011

## Problem Summary(Required) ：

`cast(cast(id_date as string) as boolean)` will optimize to `cast(id_date as boolean)`

```
mysql>    select cast(cast(cast('2020-02-02' AS date) as string) as boolean) ;
+---------------------------------------------------------------------+
| CAST(CAST(CAST('2020-02-02' AS DATE) AS VARCHAR(65533)) AS BOOLEAN) |
+---------------------------------------------------------------------+
|                                                                   1 |
+---------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql>    select cast(cast('2020-02-02' as  string) as boolean) ;
+-------------------------------------------------------+
| CAST(CAST('2020-02-02' AS VARCHAR(65533)) AS BOOLEAN) |
+-------------------------------------------------------+
|                                                  NULL |
+-------------------------------------------------------+
1 row in set (0.01 sec)
```

### CODE REVIEW

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java|3|3|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 1]
```diff
@@ -110,17 +110,18 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
             return false;
         }
 
-        // decreasing cast cannot be reduced since be use operator not equal
-        // to perform cast from integral to boolean
-        // E.g. cast(cast(10000000000000 as int) as boolean)
-        if (parent.isBoolean() && childSlotSize < grandChildSlotSize) {
+        if (!(parent.isNumericType() || parent.isStringType()) ||
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:113]: Can other types benefit from it?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:113]: Maybe, but it's hard to check each type, so only support Number/String/Boolean type

**Review by Youngwb — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**